src/main/java/hudson/plugins/copyartifact/TriggeredBuildSelector.java

```diff
@@ -138,69 +142,61 @@ public boolean isUseNewest() {
-                                (result == null)
-                                || (isUseNewest() && result.getNumber() < run.getNumber())
-                                || (!isUseNewest() && result.getNumber() > run.getNumber())
-                        ) {
-                            result = run;
-                        }
-                    }
-                } else {
-                    // Figure out the parent job and do a recursive call to getBuild
-                    Jenkins jenkins = Jenkins.getInstance();
-                    if (jenkins == null) {
-                        // to suppress findbugs warnings.
-                        LOGGER.log(
-                                Level.SEVERE,
-                                "Jenkins instance is unavailable and cannot perform copyartifact from {0} for {1}",
-                                new Object[] {
-                                        job.getFullName(),
-                                        parent.getDisplayName(),
-                                }
-                        );
-                        return null;
-                    }
-                    Job<?,?> parentJob = jenkins.getItemByFullName(upstreamProject, Job.class);
-                    if (parentJob == null) {
-                        LOGGER.log(Level.WARNING, "Upstream project doesn't exist (may be removed): {0}", upstreamProject);
-                        continue;
-                    }
-                    if (parentJob.getBuildByNumber(upstreamBuild) == null) {
-                        LOGGER.log(Level.WARNING, "Upstream build doesn't exist (may be removed): {0} #{1}", new Object[]{upstreamProject, upstreamBuild});
-                        continue;
-                    }
-                    Run<?,?> run = getBuild(
-                        job,
-                        env,
-                        filter,
-                        parentJob.getBuildByNumber(upstreamBuild));
-                    if (run != null && filter.isSelectable(run, env)) {
-                        if (
-                                (result == null)
-                                || (isUseNewest() && result.getNumber() < run.getNumber())
-                                || (!isUseNewest() && result.getNumber() > run.getNumber())
-                        ) {
-                            result = run;
-                        }
-                    }
+                Run<?, ?> upstreamRun = upstream.getUpstreamRun();
+                if (upstreamRun != null) {
+                    upstreamBuilds.add(upstreamRun);
+                }
+            }
+        }
+
+        if (isAllowUpstreamDependencies() && (parent instanceof AbstractBuild)) {
+            AbstractBuild<?, ?> parentBuild = (AbstractBuild<?,?>)parent;
+            
+            Map<AbstractProject, Integer> parentUpstreamBuilds = parentBuild.getUpstreamBuilds();
+            for (AbstractProject project : parentUpstreamBuilds.keySet()) {
```

Comment:
Please use `Map#enteySet`.

You can use findbugs to see those warnings.
Run `mvn findbugs:findbugs` after compiling your codes.